dmipy/optimizers/amico_cvxpy.py

```diff
@@ -0,0 +1,143 @@
+        Contains Dmipy multi-compartment model information.
+    x0_vector: Array of size (Nx),
+        Vector containing probability distributions of the parameters that are
+        being estimated
+    lambda_1: Array of size (Nc)
+        Vector containing L2 regularization weights for each compartment
+    lambda_2: Array of size (Nc)
+        Vector containing L1 regularization weights for each compartment
+    Nt: int
+        Integer equal to the number of equidistant sampling points over each
+        parameter range used to create atoms of observation matrix M
+
+    References
+    ----------
+    .. [1] Daducci, Alessandro, et al. "Accelerated microstructure imaging
+        via convex optimization (AMICO) from diffusion MRI data."
+        NeuroImage 105 (2015): 32-44.
+
+    .. [2] Diamond, Steven, and Stephen Boyd. "CVXPY: A Python-embedded
+        modeling language for convex optimization." The Journal of Machine
+        Learning Research 17.1 (2016): 2909-2913.
+    """
+
+    def __init__(self, acquisition_scheme, model, x0_vector=None,
+                 lambda_1=None, lambda_2=None, Nt=10):
+        self.model = model
+        self.acquisition_scheme = acquisition_scheme
+
+        if len(lambda_1) != self.model.N_models or \
+                len(lambda_2) != self.model.N_models:
+            raise ValueError("Number of regularization weights should"
+                             "correspond to the number of compartments!")
+
+        self.lambda_1 = lambda_1
+        self.lambda_2 = lambda_2
+
+        self.Nt = Nt
+
+        # Make a list of parameters that are not fixed and that require
+        # tessellation of parameter ranges
+        dir_params = [p for p in self.model.parameter_names
+                      if p.endswith('mu')]
+        grid_params =\
+            [p for p in self.model.parameter_names
+             if not p.endswith('mu') and not p.startswith('partial_volume')]
+
+        # Compute length of the vector x0
+        x0_len = 0
+        for m_idx in range(self.model.N_models):
+            m_atoms = 1
+            for p in self.model.models[m_idx].parameter_names:
+                if self.model.model_names[m_idx] + p in grid_params:
+                    m_atoms *= Nt
+            x0_len += m_atoms
+
+        # Creating parameter tessellation grids and corresponding indices
+        self.grids, self.idx = {}, {}
+        for m_idx in range(self.model.N_models):
+            model = self.model.models[m_idx]
+            model_name = self.model.model_names[m_idx]
+
+            param_sampling, grid_params_names = [], []
+            m_atoms = 1
+            for p in model.parameter_names:
+                if model_name + p not in grid_params:
+                    continue
+                grid_params_names.append(model_name + p)
+                p_range = self.model.parameter_ranges[model_name + p]
+                self.grids[model_name + p] = np.mean(p_range)
```

Comment:
I don't understand why `self.grids` for each parameter is first instantiated with the mean of the defualt parameter ranges, then in line 103 you add an array of zeros to it of size x0_len, and then finally in 104 you put the actual grid values.

why not just to the last step in 1 go and avoid extra assigning of keys?

Reply:
I've put mean because it should be some value from the range, but it doesn't matter which one (I've could have put the lowest value as well). These would be the values of parameters when we are generating atoms that belong to the other models - (for example if we are generating atom that corresponds to an isotropic signal, all parameters that belong to anisotropic signal would be mean of the corresponding range, partial volume for isotropic would be 1 and for anisotropic would be 0). In line 104, actual grid values are put only for the atoms that correspond to that model. But it is true that I could have put self.grids[model_name + p] = np.mean(p_range) + np.zeros(x0_len) in the line 94 and remove it from the line 103. 

I don't see where is the extra assigning of keys? 
Like this we can generate matrix M at once, otherwise we can go trough models and have grids for each model separately, but still when we simulate signals for matrix M, we would need to initialize parameters for the other models.

Reply:
what i meant that it looked like you are re-assigning the same keys several times, sorry for being unclear in my words!

As for the code update you proposed, I suggest you instead change it to `self.grids[model_name + p] = np.full(x0_len, p_range)` to avoid adding a constant to a zeros array, but just generate the constant array directly :-)

Reply:
Ok, thanks! :)